src/Microsoft.DotNet.Darc/src/DarcLib/LocalGitClient.cs

```diff
@@ -257,9 +257,172 @@ public Task<GitDiff> GitDiffAsync(string repoUri, string baseVersion, string tar
+                                        LibGit2Sharp.Commands.Checkout(localRepo, resolvedReference, checkoutOptions);
+                                    }
+                                }
+
+                                if (force)
+                                {
+                                    CleanRepoAndSubmodules(localRepo, _logger);
+                                }
+                            }
+                            catch   // Most likely network exception, could also be no remotes.  We can't do anything about any error here.
+                            {
+                                _logger.LogError($"After fetch, still couldn't find commit or treeish {commit} in {repoDir}.  Are you offline or missing a remote?");
+                                throw;
+                            }
+                        }
+                    }
+                }
+                catch (Exception exc)
+                {
+                    throw new Exception($"Something went wrong when checking out {commit} in {repoDir}", exc);
+                }
+            }
+        }
+
+        private static void CleanRepoAndSubmodules(LibGit2Sharp.Repository repo, ILogger log)
+        {
+            log.LogDebug($"Beginning clean of {repo.Info.WorkingDirectory} and {repo.Submodules.Count()} submodules");
+
+            LibGit2Sharp.StatusOptions options = new LibGit2Sharp.StatusOptions
+            {
+                IncludeUntracked = true,
+                RecurseUntrackedDirs = true,
+            };
+            int count = 0;
+            foreach (LibGit2Sharp.StatusEntry item in repo.RetrieveStatus(options))
+            {
+                if (item.State == LibGit2Sharp.FileStatus.NewInWorkdir)
+                {
+                    File.Delete(Path.Combine(repo.Info.WorkingDirectory, item.FilePath));
+                    ++count;
+                }
+            }
+            log.LogDebug($"Deleted {count} untracked files");
+
+            foreach (LibGit2Sharp.Submodule sub in repo.Submodules)
+            {
+                string normalizedSubPath = sub.Path.Replace('\\', Path.DirectorySeparatorChar).Replace('/', Path.DirectorySeparatorChar);
+                string subRepoPath = Path.Combine(repo.Info.WorkingDirectory, normalizedSubPath);
+                string subRepoGitFilePath = Path.Combine(subRepoPath, ".git");
+                if (!File.Exists(subRepoGitFilePath))
+                {
+                    log.LogDebug($"Submodule {sub.Name} in {subRepoPath} does not appear to be initialized, attempting to initialize now.");
+                    // hasn't been initialized yet, can happen when differest hashes have new or moved submodules
+                    repo.Submodules.Update(sub.Name, new LibGit2Sharp.SubmoduleUpdateOptions { Init = true });
+                }
+
+                log.LogDebug($"Beginning clean of submodule {sub.Name}");
```

Comment:
>log.LogDebug [](start = 16, length = 12)

log.LogDebug

Reply:
Added.

---
In reply to: [275518244](https://github.com/dotnet/arcade-services/pull/326#discussion_r275518244) [](ancestors = 275518244)

Reply:
Sweet! I pasted the incorrect piece of code but you still got the idea!

---
In reply to: [275572174](https://github.com/dotnet/arcade-services/pull/326#discussion_r275572174) [](ancestors = 275572174,275518244)